src/types.rs

```diff
@@ -30,13 +32,21 @@ impl TypeRef {
     pub fn new(id: TypeId, type_store: Rc<TypeStore>) -> Self {
         Self { id, type_store }
     }
+
+    pub fn validate(&self, value: OwnedElement, issues: &mut impl Violations) {
+        let type_def = self.type_store.get_type_by_id(self.id).unwrap();
+        for constraint in type_def.constraints() {
+            constraint.validate(value.to_owned(), issues, &self.type_store);
```

Comment:
Why does `validate` consume `value`? (Said another way: why can't `value` be a `&OwnedElement` instead of an `OwnedElement`?) We shouldn't need to clone the element (with `to_owned()`) each time we recurse.

Reply:
Yes I will change this to be `&OwnedElement` instead.